File: functions/update-whatsapp-settings/whatsapp_service.py
import os
import json
import requests
import logging
import time
import re
from typing import Dict, Any, Optional, List, Tuple
from datetime import datetime, timedelta

logger = logging.getLogger(__name__)

class WhatsAppError(Exception):
    """Custom exception for WhatsApp-related errors"""
    def __init__(self, message: str, error_code: Optional[str] = None, retryable: bool = False):
        self.message = message
        self.error_code = error_code
        self.retryable = retryable
        super().__init__(self.message)

class GreenAPIClient:
    """Green API client for WhatsApp messaging"""
    
    def __init__(self, instance_id: str, token: str):
        self.instance_id = instance_id
        self.token = token
        self.base_url = f"https://api.green-api.com/waInstance{instance_id}"
        self.session = requests.Session()
        self.session.timeout = 30
        
    def send_message(self, phone_number: str, message: str) -> Dict[str, Any]:
        """
        Send WhatsApp message via Green API
        
        Args:
            phone_number: Phone number in international format
            message: Message text to send
            
        Returns:
            API response dictionary
            
        Raises:
            WhatsAppError: If message sending fails
        """
        try:
            # Format phone number
            formatted_phone = self._format_phone_number(phone_number)
            
            # Prepare request
            url = f"{self.base_url}/sendMessage/{self.token}"
            payload = {
                "chatId": f"{formatted_phone}@c.us",
                "message": message
            }
            
            logger.info(f"Sending WhatsApp message to {formatted_phone[:5]}****")
            
            response = self.session.post(url, json=payload)
            
            if response.status_code == 200:
                result = response.json()
                if result.get('idMessage'):
                    logger.info(f"Message sent successfully: {result.get('idMessage')}")
                    return result
                else:
                    raise WhatsAppError(
                        f"Message sending failed: {result.get('error', 'Unknown error')}",
                        error_code="SEND_FAILED",
                        retryable=True
                    )
            elif response.status_code == 429:
                raise WhatsAppError(
                    "Rate limit exceeded",
                    error_code="RATE_LIMIT",
                    retryable=True
                )
            elif response.status_code == 401:
                raise WhatsAppError(
                    "Invalid Green API credentials",
                    error_code="AUTH_FAILED",
                    retryable=False
                )
            else:
                raise WhatsAppError(
                    f"HTTP error {response.status_code}: {response.text}",
                    error_code="HTTP_ERROR",
                    retryable=True
                )
                
        except requests.exceptions.Timeout:
            raise WhatsAppError(
                "Request timeout",
                error_code="TIMEOUT",
                retryable=True
            )
        except requests.exceptions.ConnectionError:
            raise WhatsAppError(
                "Connection error",
                error_code="CONNECTION_ERROR",
                retryable=True
            )
        except Exception as e:
            if isinstance(e, WhatsAppError):
                raise
            raise WhatsAppError(
                f"Unexpected error: {str(e)}",
                error_code="UNKNOWN_ERROR",
                retryable=False
            )
    
    def test_connection(self) -> Tuple[bool, Optional[str]]:
        """
        Test Green API connection and credentials
        
        Returns:
            Tuple of (success, error_message)
        """
        try:
            url = f"{self.base_url}/getSettings/{self.token}"
            response = self.session.get(url)
            
            if response.status_code == 200:
                return True, None
            elif response.status_code == 401:
                return False, "Invalid credentials"
            else:
                return False, f"Connection test failed: {response.status_code}"
                
        except Exception as e:
            return False, f"Connection error: {str(e)}"
    
    def _format_phone_number(self, phone_number: str) -> str:
        """
        Format phone number for Green API
        
        Args:
            phone_number: Raw phone number
            
        Returns:
            Formatted phone number without + prefix
        """
        # Remove all non-digit characters
        digits_only = re.sub(r'\D', '', phone_number)
        
        # Remove leading + if present
        if digits_only.startswith('7') and len(digits_only) == 11:
            # Russian number format
            return digits_only
        elif digits_only.startswith('8') and len(digits_only) == 11:
            # Convert Russian 8 to 7
            return '7' + digits_only[1:]
        elif len(digits_only) >= 10:
            # International format
            return digits_only
        else:
            raise WhatsAppError(
                f"Invalid phone number format: {phone_number}",
                error_code="INVALID_PHONE",
                retryable=False
            )

class MessageTemplateProcessor:
    """Processes message templates with variable substitution"""
    
    # Default templates
    DEFAULT_TEMPLATES = {
        'reminder_7_days': "Здравствуйте, {client_name}! Напоминаем, что ваш платеж по рассрочке в размере {installment_amount} руб. за {product_name} должен быть внесен через {days_remaining} дней ({due_date}). Пожалуйста, подготовьте средства для оплаты.",
        'reminder_due_today': "Здравствуйте, {client_name}! Ваш платеж по рассрочке в размере {installment_amount} руб. за {product_name} должен быть внесен сегодня ({due_date}). Пожалуйста, произведите оплату.",
        'reminder_manual': "Здравствуйте, {client_name}! Напоминаем о вашем платеже по рассрочке в размере {installment_amount} руб. за {product_name}. Дата платежа: {due_date}."
    }
    
    @staticmethod
    def process_template(template: str, variables: Dict[str, Any]) -> str:
        """
        Process template with variable substitution
        
        Args:
            template: Template string with {variable} placeholders
            variables: Dictionary of variable values
            
        Returns:
            Processed message string
        """
        try:
            # Ensure all variables are strings
            str_variables = {k: str(v) if v is not None else '' for k, v in variables.items()}
            
            # Replace variables in template
            processed = template.format(**str_variables)
            
            logger.debug(f"Template processed successfully")
            return processed
            
        except KeyError as e:
            logger.error(f"Missing template variable: {e}")
            raise WhatsAppError(
                f"Template processing failed: missing variable {e}",
                error_code="TEMPLATE_ERROR",
                retryable=False
            )
        except Exception as e:
            logger.error(f"Template processing error: {e}")
            raise WhatsAppError(
                f"Template processing failed: {str(e)}",
                error_code="TEMPLATE_ERROR",
                retryable=False
            )
    
    @staticmethod
    def validate_template(template: str) -> Tuple[bool, List[str]]:
        """
        Validate template format and extract variables
        
        Args:
            template: Template string to validate
            
        Returns:
            Tuple of (is_valid, list_of_variables)
        """
        try:
            # Extract variables from template
            import string
            formatter = string.Formatter()
            variables = [field_name for _, field_name, _, _ in formatter.parse(template) if field_name]
            
            # Check for valid variable names
            valid_variables = {
                'client_name', 'installment_amount', 'due_date', 
                'days_remaining', 'product_name', 'total_amount'
            }
            
            invalid_vars = [var for var in variables if var not in valid_variables]
            
            if invalid_vars:
                logger.warning(f"Invalid template variables: {invalid_vars}")
                return False, invalid_vars
            
            return True, variables
            
        except Exception as e:
            logger.error(f"Template validation error: {e}")
            return False, [str(e)]

class WhatsAppService:
    """Main WhatsApp service class"""
    
    def __init__(self, instance_id: str, token: str):
        self.client = GreenAPIClient(instance_id, token)
        self.template_processor = MessageTemplateProcessor()
    
    def send_reminder(self, phone_number: str, template: str, variables: Dict[str, Any], max_retries: int = 3) -> Dict[str, Any]:
        """
        Send WhatsApp reminder with retry logic
        
        Args:
            phone_number: Client phone number
            template: Message template
            variables: Template variables
            max_retries: Maximum retry attempts
            
        Returns:
            Result dictionary with status and details
        """
        result = {
            'status': 'failed',
            'message_id': None,
            'error': None,
            'attempts': 0
        }
        
        try:
            # Process template
            message = self.template_processor.process_template(template, variables)
            
            # Attempt to send with retries
            for attempt in range(max_retries):
                result['attempts'] = attempt + 1
                
                try:
                    response = self.client.send_message(phone_number, message)
                    result['status'] = 'success'
                    result['message_id'] = response.get('idMessage')
                    return result
                    
                except WhatsAppError as e:
                    result['error'] = e.message
                    
                    if not e.retryable or attempt == max_retries - 1:
                        break
                    
                    # Exponential backoff
                    wait_time = 2 ** attempt
                    logger.info(f"Retrying in {wait_time} seconds (attempt {attempt + 1}/{max_retries})")
                    time.sleep(wait_time)
            
            return result
            
        except Exception as e:
            result['error'] = str(e)
            return result
    
    def test_connection(self) -> Tuple[bool, Optional[str]]:
        """Test Green API connection"""
        return self.client.test_connection()

def create_whatsapp_service(instance_id: str, token: str) -> WhatsAppService:
    """Factory function to create WhatsApp service instance"""
    return WhatsAppService(instance_id, token)

File: lib/core/services/whatsapp_service.dart
import 'dart:convert';
import 'dart:developer' as developer;
import 'dart:math';
import 'package:http/http.dart' as http;

/// Green API client wrapper for WhatsApp messaging
class WhatsAppService {
  static const String _baseUrl = 'https://api.green-api.com';
  
  final String instanceId;
  final String token;
  final http.Client _httpClient;
  
  // Rate limiting properties
  static const int _maxRequestsPerMinute = 30; // Green API typical limit
  static const Duration _rateLimitWindow = Duration(minutes: 1);
  static final Map<String, List<DateTime>> _requestHistory = {};
  static final Map<String, DateTime> _lastRateLimitReset = {};

  WhatsAppService({
    required this.instanceId,
    required this.token,
    http.Client? httpClient,
  }) : _httpClient = httpClient ?? http.Client();

  /// Get the base URL for Green API requests
  String get _apiUrl => '$_baseUrl/waInstance$instanceId';

  /// Get common headers for Green API requests
  Map<String, String> get _headers => {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $token',
      };

  /// Send a text message via Green API with retry logic
  Future<Map<String, dynamic>> sendTextMessage({
    required String phoneNumber,
    required String message,
    int maxRetries = 3,
  }) async {
    // Validate phone number
    if (!isValidPhoneNumber(phoneNumber)) {
      throw WhatsAppException(
        'Invalid phone number format: $phoneNumber',
        retryable: false,
      );
    }

    return await _executeWithRetry(
      () => _sendTextMessageInternal(phoneNumber, message),
      maxRetries: maxRetries,
      operation: 'sendTextMessage',
    );
  }

  /// Send a templated message with variable substitution
  Future<Map<String, dynamic>> sendTemplatedMessage({
    required String phoneNumber,
    required String template,
    required Map<String, String> variables,
    int maxRetries = 3,
  }) async {
    // Process template with variables
    final processedMessage = processMessageTemplate(template, variables);
    
    return await sendTextMessage(
      phoneNumber: phoneNumber,
      message: processedMessage,
      maxRetries: maxRetries,
    );
  }

  /// Internal method to send text message (without retry logic)
  Future<Map<String, dynamic>> _sendTextMessageInternal(
    String phoneNumber,
    String message,
  ) async {
    // Check rate limiting before making request
    await _checkRateLimit();
    
    _logRequest('sendTextMessage', {
      'phoneNumber': phoneNumber,
      'message': message.length > 50 ? '${message.substring(0, 50)}...' : message,
    });

    final url = '$_apiUrl/sendMessage/$token';
    final body = {
      'chatId': _formatPhoneNumber(phoneNumber),
      'message': message,
    };

    final response = await _httpClient.post(
      Uri.parse(url),
      headers: _headers,
      body: json.encode(body),
    );

    // Record this request for rate limiting
    _recordRequest();

    final result = await _handleResponse(response);
    
    _logResponse('sendTextMessage', result);
    return result;
  }

  /// Check if we're within rate limits, wait if necessary
  Future<void> _checkRateLimit() async {
    final instanceKey = instanceId;
    final now = DateTime.now();
    
    // Initialize request history for this instance if needed
    _requestHistory[instanceKey] ??= [];
    
    // Clean up old requests outside the rate limit window
    _requestHistory[instanceKey]!.removeWhere(
      (requestTime) => now.difference(requestTime) > _rateLimitWindow,
    );
    
    // Check if we're at the rate limit
    if (_requestHistory[instanceKey]!.length >= _maxRequestsPerMinute) {
      // Calculate how long to wait
      final oldestRequest = _requestHistory[instanceKey]!.first;
      final waitTime = _rateLimitWindow - now.difference(oldestRequest);
      
      if (waitTime.inMilliseconds > 0) {
        developer.log(
          'Rate limit reached. Waiting ${waitTime.inMilliseconds}ms...',
          name: 'WhatsAppService',
          level: 900, // Warning level
        );
        
        await Future.delayed(waitTime);
        
        // Clean up again after waiting
        final newNow = DateTime.now();
        _requestHistory[instanceKey]!.removeWhere(
          (requestTime) => newNow.difference(requestTime) > _rateLimitWindow,
        );
      }
    }
  }

  /// Record a request for rate limiting purposes
  void _recordRequest() {
    final instanceKey = instanceId;
    final now = DateTime.now();
    
    _requestHistory[instanceKey] ??= [];
    _requestHistory[instanceKey]!.add(now);
    _lastRateLimitReset[instanceKey] = now;
  }

  /// Get current rate limit status
  Map<String, dynamic> getRateLimitStatus() {
    final instanceKey = instanceId;
    final now = DateTime.now();
    
    _requestHistory[instanceKey] ??= [];
    
    // Clean up old requests
    _requestHistory[instanceKey]!.removeWhere(
      (requestTime) => now.difference(requestTime) > _rateLimitWindow,
    );
    
    final currentRequests = _requestHistory[instanceKey]!.length;
    final remainingRequests = _maxRequestsPerMinute - currentRequests;
    
    DateTime? resetTime;
    if (_requestHistory[instanceKey]!.isNotEmpty) {
      final oldestRequest = _requestHistory[instanceKey]!.first;
      resetTime = oldestRequest.add(_rateLimitWindow);
    }
    
    return {
      'current_requests': currentRequests,
      'max_requests': _maxRequestsPerMinute,
      'remaining_requests': remainingRequests,
      'window_duration_minutes': _rateLimitWindow.inMinutes,
      'reset_time': resetTime?.toIso8601String(),
      'is_rate_limited': remainingRequests <= 0,
    };
  }

  /// Process message template with variable substitution
  static String processMessageTemplate(
    String template,
    Map<String, String> variables,
  ) {
    String processedMessage = template;
    
    // Replace each variable in the template
    variables.forEach((key, value) {
      // Support both {variable} and {{variable}} formats
      processedMessage = processedMessage.replaceAll('{$key}', value);
      processedMessage = processedMessage.replaceAll('{{$key}}', value);
    });
    
    return processedMessage;
  }

  /// Validate phone number format
  static bool isValidPhoneNumber(String phoneNumber) {
    // Remove all non-digit characters for validation
    final cleaned = phoneNumber.replaceAll(RegExp(r'[^\d]'), '');
    
    // Check if it's a valid length (7-15 digits as per E.164 standard)
    if (cleaned.length < 7 || cleaned.length > 15) {
      return false;
    }
    
    // Check if it contains only digits
    return RegExp(r'^\d+$').hasMatch(cleaned);
  }

  /// Execute operation with retry logic and exponential backoff
  Future<T> _executeWithRetry<T>(
    Future<T> Function() operation, {
    required int maxRetries,
    required String operationName,
  }) async {
    int attempt = 0;
    
    while (attempt < maxRetries) {
      try {
        return await operation();
      } catch (e) {
        attempt++;
        
        if (e is WhatsAppException && !e.retryable) {
          // Don't retry non-retryable errors
          rethrow;
        }
        
        if (attempt >= maxRetries) {
          // Max retries reached
          _logError('$operationName (final attempt)', e);
          rethrow;
        }
        
        // Calculate exponential backoff delay
        final delaySeconds = pow(2, attempt - 1).toInt();
        final jitter = Random().nextInt(1000); // Add jitter to prevent thundering herd
        final totalDelay = Duration(seconds: delaySeconds, milliseconds: jitter);
        
        _logError('$operationName (attempt $attempt/$maxRetries)', e);
        developer.log(
          'Retrying in ${totalDelay.inMilliseconds}ms...',
          name: 'WhatsAppService',
          level: 900, // Warning level
        );
        
        await Future.delayed(totalDelay);
      }
    }
    
    throw WhatsAppException('Max retries exceeded for $operationName');
  }

  /// Test the connection to Green API
  Future<Map<String, dynamic>> testConnection() async {
    try {
      _logRequest('testConnection', {'instanceId': instanceId});

      final url = '$_apiUrl/getSettings/$token';
      
      final response = await _httpClient.get(
        Uri.parse(url),
        headers: _headers,
      );

      final result = await _handleResponse(response);
      
      _logResponse('testConnection', result);
      return result;
    } catch (e) {
      _logError('testConnection', e);
      rethrow;
    }
  }

  /// Get account information from Green API
  Future<Map<String, dynamic>> getAccountInfo() async {
    try {
      _logRequest('getAccountInfo', {'instanceId': instanceId});

      final url = '$_apiUrl/getWaSettings/$token';
      
      final response = await _httpClient.get(
        Uri.parse(url),
        headers: _headers,
      );

      final result = await _handleResponse(response);
      
      _logResponse('getAccountInfo', result);
      return result;
    } catch (e) {
      _logError('getAccountInfo', e);
      rethrow;
    }
  }

  /// Format phone number for Green API (must include country code and @c.us suffix)
  String _formatPhoneNumber(String phoneNumber) {
    // Remove all non-digit characters
    String cleaned = phoneNumber.replaceAll(RegExp(r'[^\d]'), '');
    
    // Add country code if not present (assuming +7 for Russian numbers)
    if (!cleaned.startsWith('7') && cleaned.length == 10) {
      cleaned = '7$cleaned';
    }
    
    // Add @c.us suffix for Green API
    return '$cleaned@c.us';
  }

  /// Handle HTTP response and parse JSON
  Future<Map<String, dynamic>> _handleResponse(http.Response response) async {
    final responseBody = response.body;
    
    if (response.statusCode >= 200 && response.statusCode < 300) {
      try {
        return json.decode(responseBody) as Map<String, dynamic>;
      } catch (e) {
        throw WhatsAppException(
          'Failed to parse response JSON: $e',
          statusCode: response.statusCode,
          retryable: false,
        );
      }
    } else {
      // Try to parse error response
      Map<String, dynamic>? errorData;
      try {
        errorData = json.decode(responseBody) as Map<String, dynamic>;
      } catch (_) {
        // Ignore JSON parsing errors for error responses
      }

      final errorMessage = errorData?['error'] ?? 
                          errorData?['message'] ?? 
                          'HTTP ${response.statusCode}: ${response.reasonPhrase}';

      throw WhatsAppException(
        errorMessage,
        statusCode: response.statusCode,
        retryable: _isRetryableError(response.statusCode),
        errorData: errorData,
      );
    }
  }

  /// Determine if an error is retryable based on status code
  bool _isRetryableError(int statusCode) {
    return statusCode >= 500 || // Server errors
           statusCode == 429 || // Rate limiting
           statusCode == 408;   // Request timeout
  }

  /// Log API request for debugging
  void _logRequest(String method, Map<String, dynamic> params) {
    developer.log(
      'WhatsApp API Request: $method',
      name: 'WhatsAppService',
      level: 800, // Info level
      error: null,
      stackTrace: null,
      zone: null,
      time: DateTime.now(),
    );
    
    developer.log(
      'Request params: ${json.encode(params)}',
      name: 'WhatsAppService',
      level: 700, // Debug level
    );
  }

  /// Log API response for debugging
  void _logResponse(String method, Map<String, dynamic> response) {
    developer.log(
      'WhatsApp API Response: $method',
      name: 'WhatsAppService',
      level: 800, // Info level
    );
    
    developer.log(
      'Response data: ${json.encode(response)}',
      name: 'WhatsAppService',
      level: 700, // Debug level
    );
  }

  /// Log API error for debugging
  void _logError(String method, dynamic error) {
    developer.log(
      'WhatsApp API Error: $method',
      name: 'WhatsAppService',
      level: 1000, // Error level
      error: error,
      stackTrace: error is Error ? error.stackTrace : null,
    );
  }

  /// Dispose of resources
  void dispose() {
    _httpClient.close();
  }
}

/// Custom exception for WhatsApp API errors
class WhatsAppException implements Exception {
  final String message;
  final int? statusCode;
  final bool retryable;
  final Map<String, dynamic>? errorData;

  const WhatsAppException(
    this.message, {
    this.statusCode,
    this.retryable = false,
    this.errorData,
  });

  @override
  String toString() {
    return 'WhatsAppException: $message${statusCode != null ? ' (HTTP $statusCode)' : ''}';
  }
}